### PR TITLE
Add npm-react-codemod

### PR DESCRIPTION
**Pull request by cpojer**
This adds `react-codemod` with the following codemods:

* `getDOMNode` to `React.findDOMNode`
* `PureRenderMixin` to `shouldComponentUpdate` using `shallowCompare`
* `React.createClass` to ES6 classes.

See the [README](https://github.com/facebook/react/blob/de6d26fc87c59319329a2cb5dedeaf4c82501dfc/npm-react-codemod/README.md) for a detailed explanation of how the codemods work.

The codemods are based on [jscodeshift](https://github.com/facebook/jscodeshift) and the project structure is based on [js-codemod](https://github.com/cpojer/js-codemod).

The tests can be run using `npm test`. Publishing this to npm should work as I have fiddled around with `js-codemod` enough.

@zpao can we pull in https://github.com/facebook/react/pull/3355 for React 0.13.2? I thought it was going into 0.13.1 but I guess it didn't :)

### CODE REVIEW

File: npm-react-codemod/README.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    React after the constructor has been run.
+  * Changes `return StateObject` from `getInitialState` to assign `this.state`
+    directly.
+
+### Recast Options
+
+Options to [recast](https://github.com/benjamn/recast)'s printer can be provided
+through the `printOptions` command line argument
+
+ * `react-codemod use-strict <file> --printOptions='{"quote":"double"}'`
```
Comment (zpao) [inline npm-react-codemod/README.md]: use-strict doesn't exist, perhaps use a real example

**Comment by zpao** [comment]
Lots of lint to fix :) You might need a custom `.eslintrc` in your folder. `grunt lint` is a bit noisy for warnings. You can also run `npm run lint npm-react-codemod` (which will only run for `src/` and your folder).

File: npm-react-codemod/scripts/test-preprocess.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+"use strict";
+
+var reactTools = require('react-tools');
+
+function process(source) {
+    return reactTools.transform(source, {harmony: true, stripTypes: true});
```
Comment (zpao) [inline npm-react-codemod/scripts/test-preprocess.js]: 4 space indent? you monster.

File: npm-react-codemod/react-codemod [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+DIR=$(npm root -g)/react-codemod
+TRANSFORM=$1
+shift
+
+$DIR/node_modules/.bin/jscodeshift -t $DIR/build/$TRANSFORM.js $@
```
Comment (zpao) [inline npm-react-codemod/react-codemod]: I'm pretty sure this isn't going to work on Windows. I thought we were going to make this a js script that used node to shell out to jscodeshift?

Comment (cpojer) [inline npm-react-codemod/react-codemod]: meh, that's no fun. Let me see what I can do.

**Comment by zpao** [comment]
fun fact, I can't actually use `react-codemod` in node 0.10 because it doesn't support `const` and those aren't transformed away.

**Comment by cpojer** [comment]
I'm on v0.10.37 and it works fine for me? Should I move back to var's?

**Comment by cpojer** [comment]
Changed to use babel.

**Comment by zpao** [comment]
ok, should be good to go

**Comment by spicyj** [comment]
jest now fails because it picks up your transform tests but jscodeshift isn't in the root package.json.

**Comment by cpojer** [comment]
oh damn, why didn't travis not catch that? :(

**Comment by zpao** [comment]
oh hmm, I thought i actually checked that locally…

> why didn't travis not catch that?
Because we're having issues with travis and jest.

**Pull request merged**